text/0000-bool-to-option.md

```diff
@@ -0,0 +1,117 @@
+common pattern into a method will make code more readable and require less repetitive typing on the
+user's part.
+
+A method for converting from `bool` to `Option<T>` has been requested several times in the past
+[[1]](https://github.com/rust-lang/rfcs/pull/2180)
+[[2]](https://github.com/rust-lang/rust/issues/50523)
+[[3]](https://github.com/rust-lang/rfcs/issues/2606) and shows a significant desire from users.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Following this proposal, we will add two methods to `bool`:
+
+```rust
+impl bool {
+    fn then_some<T>(self, t: T) -> Option<T> {
+        if self {
+            Some(t)
+        } else {
+            None
+        }
+    }
+
+    fn then<T, F: FnOnce() -> T>(self, f: F) -> Option<T> {
+        if self {
+            Some(f())
+        } else {
+            None
+        }
+    }
+}
+```
+
+The primitive type `bool` currently has no methods, so it will be necessary to add support similarly
+to other primitive types that do have methods, like
+[`char`](https://doc.rust-lang.org/src/core/char/methods.rs.html#11-1393). This will require the
+addition of a new lang item: `#[lang = "bool"]`.
+
+A reference implementation is provided in [#64255](https://github.com/rust-lang/rust/pull/64255).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Save the usual drawbacks of adding a new method to the standard library, there are no
+drawbacks.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The implementations here are the only reasonable ones.
+
+The following names have been suggested in the past. The choice in this proposal has been made to
+avoid possible confusion (e.g. avoiding a name that suggests a conversion from `bool` to
+`Option<()>` rather than `Option<T>`), and to be consist with existing naming conventions,
+but ultimately comes down to personal preference and indication of
+[community consensus](https://github.com/rust-lang/rfcs/pull/2757#issuecomment-544437510) following
+the [examples in the rustc codebase](https://github.com/rust-lang/rust/pull/65195).
+
+- [`to_option` and `to_option_with`](https://github.com/rust-lang/rfcs/pull/2757#issuecomment-529228631)
+- [`to_opt` or `then_some`](https://github.com/rust-lang/rfcs/issues/2606#issuecomment-476019577)
+- [`then` and `then_do`](https://github.com/rust-lang/rfcs/pull/2180#issuecomment-350498489)
+- [`some`](https://github.com/rust-lang/rfcs/issues/2606#issue-387773675)
+- [`as_some` and `as_some_from`](https://docs.rs/boolinator/2.4.0/boolinator/trait.Boolinator.html)
+- [`some_if` and `lazy_if`](https://github.com/rust-lang/rfcs/pull/2180)
+
+This functionality could instead be provided by a crate (e.g.
+[boolinator](https://docs.rs/boolinator/2.4.0/boolinator/)), but this functionality is commonly
+desired and an obvious candidate for the standard library, where an external crate for such simple
+functionality is not convenient.
+
```

Comment:
Might as well mention [`if some_condition { t }`](https://github.com/rust-lang/rfcs/issues/961#issuecomment-290339751) as an alternative.

Reply:
Inlining a method body is always a possible alternative, so mentioning it specifically here isn't useful.

Reply:
@varkor that's does not mean inlining the method body, but to make `if x { y }` evaluates to `Some(y)` if `x` is true and `None` if `x` is false (basically an even more controversial version of https://internals.rust-lang.org/t/pre-rfc-break-with-value-in-for-while-loops/11208)

Reply:
Ah, I see. Unless there are serious proponents of that approach, I don't think it's worth including, as it seems like an unviable alternative to me.